File: Source/INSPersistentStoreDescription.m
//
//  INSPersistentStoreDescription.m
//  INSPersistentStoreDescription
//
//  Created by Michal Zaborowski on 15.06.2016.
//  Copyright © 2016 Michal Zaborowski. All rights reserved.
//

#import "INSPersistentStoreDescription.h"
#import <objc/runtime.h>

@interface INSPersistentStoreDescription ()
@property (nonatomic, copy) NSMutableDictionary<NSString *, NSObject *> *options;
@end

@implementation INSPersistentStoreDescription
@synthesize URL = _url;

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ (type: %@, url: %@)", [super description], self.type, self.URL];
}

- (NSDictionary<NSString *, NSObject *> *)options {
    return [self->_options copy];
}

- (BOOL)shouldAddStoreAsynchronously {
    NSNumber *option = (NSNumber *)[self.options objectForKey:@"NSAddStoreAsynchronouslyOption"];
    return [option boolValue];
}

- (void)setShouldAddStoreAsynchronously:(BOOL)shouldAddStoreAsynchronously {
    [self setOption:@(shouldAddStoreAsynchronously) forKey:@"NSAddStoreAsynchronouslyOption"];
}

- (BOOL)shouldMigrateStoreAutomatically {
    NSNumber *option = (NSNumber *)[self.options objectForKey:NSMigratePersistentStoresAutomaticallyOption];
    return [option boolValue];
}

- (void)setShouldMigrateStoreAutomatically:(BOOL)shouldMigrateStoreAutomatically {
    [self setOption:@(shouldMigrateStoreAutomatically) forKey:NSMigratePersistentStoresAutomaticallyOption];
}

- (BOOL)shouldInferMappingModelAutomatically {
    NSNumber *option = (NSNumber *)[self.options objectForKey:NSInferMappingModelAutomaticallyOption];
    return [option boolValue];
}

- (void)setShouldInferMappingModelAutomatically:(BOOL)shouldInferMappingModelAutomatically {
    [self setOption:@(shouldInferMappingModelAutomatically) forKey:NSInferMappingModelAutomaticallyOption];
}

- (NSDictionary<NSString *, NSObject *> *)sqlitePragmas {
    return (NSDictionary<NSString *, NSObject *> *)[self.options objectForKey:NSSQLitePragmasOption] ?: [NSDictionary dictionary];
}

- (BOOL)isReadOnly {
    NSNumber *option = (NSNumber *)[self.options objectForKey:NSReadOnlyPersistentStoreOption];
    return [option boolValue];
}

- (void)setReadOnly:(BOOL)readOnly {
    [self setOption:@(readOnly) forKey:NSReadOnlyPersistentStoreOption];
}

- (NSTimeInterval)timeout {
    NSNumber *option = (NSNumber *)[self.options objectForKey:NSPersistentStoreTimeoutOption];
    return [option doubleValue];
}

- (void)setTimeout:(NSTimeInterval)timeout {
    [self setOption:[NSNumber numberWithDouble:timeout] forKey:NSPersistentStoreTimeoutOption];
}

+ (instancetype)persistentStoreDescriptionWithURL:(NSURL *)URL {
    return [[self alloc] initWithURL:URL];
}

- (instancetype)init {
    return [self initWithURL:[NSURL fileURLWithPath:@"/dev/null"]];
}

- (instancetype)initWithURL:(NSURL *)url {
    if (self = [super init]) {
        self->_url = [url copy];
        self->_type = @"SQLite";
        self->_options = [@{} mutableCopy];
        self.shouldInferMappingModelAutomatically = YES;
        self.shouldMigrateStoreAutomatically = YES;
    }
    return self;
}

- (void)setValue:(nullable NSObject *)value forPragmaNamed:(NSString *)name {
    NSMutableDictionary<NSString *, NSObject *> *pragmas = [self.sqlitePragmas mutableCopy];
    if (value) {
        [pragmas setObject:value forKey:name];
    } else {
        [pragmas removeObjectForKey:name];
    }
    [self setOption:[pragmas copy] forKey:NSSQLitePragmasOption];
}

- (void)setOption:(NSObject *)option forKey:(NSString *)key {
    NSMutableDictionary<NSString *, NSObject *> *options = self->_options;
    if (option) {
        [options setObject:option forKey:key];
    } else {
        [options removeObjectForKey:key];
    }
}

- (NSUInteger)hash {
    NSURL *URL = self.URL;
    if ([URL isFileURL]) {
        return [URL hash];
    }
    char *result = realpath([[URL path] UTF8String], 0x0);
    if (result) {
        return [[NSURL fileURLWithPath:[NSString stringWithUTF8String:result]] hash];
    }

    return [URL hash];
}

- (id)copyWithZone:(NSZone *)zone {
    INSPersistentStoreDescription *copy = [[INSPersistentStoreDescription alloc] initWithURL:self->_url];
    copy.configuration = [self->_configuration copy];
    copy.type = [self->_type copy];
    copy.options = [self->_options mutableCopy];
    return copy;
}

@end

#if NS_PERSISTENT_STORE_NOT_AVAILABLE_IN_SDK
@implementation NSPersistentStoreDescription @end
#endif

File: Source/Swift/INSPersistentStoreDescription.swift
//
//  INSPersistentStoreDescription.swift
//  INSPersistentContainer-Swift2
//
//  Created by Michal Zaborowski on 24.06.2016.
//  Copyright © 2016 Inspace Labs Sp z o. o. Spółka Komandytowa. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.

import Foundation
import CoreData

// An instance of NSPersistentStoreDescription encapsulates all information needed to describe a persistent store.

open class INSPersistentStoreDescription: CustomStringConvertible {
    open var type: String = "SQLite"
    open var configuration: String?
    
    open var url: Foundation.URL
    open private(set) var options: [String : Any] = [:]

    open var sqlitePragmas: [String : Any] {
        return options[NSSQLitePragmasOption] as? [String : Any] ?? [:]
    }
    
    open var description: String {
        return "type: \(type), url: \(url)"
    }
    
    // addPersistentStore-time behaviours
    open var shouldAddStoreAsynchronously: Bool {
        set {
            options["NSAddStoreAsynchronouslyOption"] = newValue
        }
        get {
            return options["NSAddStoreAsynchronouslyOption"] as? Bool ?? false
        }
    }
    open var shouldMigrateStoreAutomatically: Bool {
        set {
            options[NSMigratePersistentStoresAutomaticallyOption] = newValue
        }
        get {
            return options[NSMigratePersistentStoresAutomaticallyOption] as? Bool ?? false
        }
    }
    open var shouldInferMappingModelAutomatically: Bool {
        set {
            options[NSInferMappingModelAutomaticallyOption] = newValue
        }
        get {
            return options[NSInferMappingModelAutomaticallyOption] as? Bool ?? false
        }
    }
    
    // Store options
    open var isReadOnly: Bool {
        set {
            options[NSReadOnlyPersistentStoreOption] = newValue
        }
        get {
            return options[NSReadOnlyPersistentStoreOption] as? Bool ?? false
        }
    }
    open var timeout: TimeInterval {
        set {
            options[NSPersistentStoreTimeoutOption] = newValue
        }
        get {
            return options[NSPersistentStoreTimeoutOption] as? TimeInterval ?? 0
        }
    }
    
    // Returns a store description instance with default values for the store located at `URL` that can be used immediately with `addPersistentStoreWithDescription:completionHandler:`.
    public init(url: Foundation.URL) {
        self.url = (url as NSURL).copy() as! Foundation.URL
        self.shouldMigrateStoreAutomatically = true
        self.shouldInferMappingModelAutomatically = true
    }
    
    open func setValue(_ value: NSObject?, forPragmaNamed name: String) {
        var pragmas = sqlitePragmas
        if let value = value {
            pragmas[name] = value
        } else {
            pragmas.removeValue(forKey: name)
        }
        setOption(pragmas as NSObject?, forKey: NSSQLitePragmasOption)
    }
    
    open func setOption(_ option: NSObject?, forKey key: String) {
        var options = self.options
        if let option = option {
            options[key] = option
        } else {
            options.removeValue(forKey: key)
        }
    }
}

extension NSPersistentStoreCoordinator {
    open func ins_addPersistentStore(with storeDescription: INSPersistentStoreDescription, completionHandler block: @escaping (INSPersistentStoreDescription, Error?) -> Swift.Void) {
        if storeDescription.shouldAddStoreAsynchronously {
            DispatchQueue.global(qos: .background).async(execute: {
                do {
                    try self.addPersistentStore(ofType: storeDescription.type, configurationName: storeDescription.configuration, at: storeDescription.url, options: storeDescription.options)
                    block(storeDescription, nil)
                } catch let error as NSError {
                    block(storeDescription, error)
                }
            })
        } else {
            do {
                try self.addPersistentStore(ofType: storeDescription.type, configurationName: storeDescription.configuration, at: storeDescription.url, options: storeDescription.options)
                block(storeDescription, nil)
            } catch let error as NSError {
                block(storeDescription, error)
            }
        }
    }
}
